src/modules/joinOrderOptimization.js

```diff
@@ -0,0 +1,460 @@
+        right: rightJoinExpr,
+        type: 'inner',
+        condition: formatJoinConditions(joinConditions),
+      },
+    };
+  }
+  return leftJoinExpr;
+};
+
+/**
+ * Gets the relational expression for the given join order
+ * @param graph - the relational graph
+ * @param bestJoinOrder - the best join order
+ * @param globalSelections - the global selections
+ */
+const getRelationalExpression = (
+  graph: Graph,
+  bestJoinOrder: Array<string | Array<JoinCondition>>,
+  globalSelections: Array<SelectionCondition>
+): {[key: string]: any} => {
+  const tablesWithSelections: {[key: string]: any} = {};
+  for (const tableName in graph) {
+    const selections = graph[tableName]?.selections ?? [];
+    if (selections.length > 0) {
+      tablesWithSelections[tableName] = getSelectionExpression(
+        tableName,
+        selections
+      );
+    }
+  }
+
+  const joinExpr = parseJoinOrderExpression(
+    tablesWithSelections,
+    bestJoinOrder
+  );
+  return getSelectionExpression(joinExpr, globalSelections);
+};
+
+/**
+ * Performs join order optimization
+ * @param graph - the relational graph
+ * @param globalSelections - the global selections
+ * @return the relational expression
+ */
+const joinOrderOptimization = (
+  graph: Graph,
+  globalSelections: Array<SelectionCondition>
+): {[key: string]: any} => {
+  const queue = new PriorityQueue();
+  // initialize the queue with all the nodes
+  for (const node in graph) {
+    queue.enqueue(
+      new JoinOrderQueueElement(graph, [node], getTableData(node), 0, [node])
+    );
+  }
+
+  let bestCost = Number.MAX_SAFE_INTEGER;
+  let bestJoinTables: Array<string> = [];
+  let bestJoinOrder: Array<string | Array<JoinCondition>> = [];
+  const JOIN_ORDER_SIZE = Object.keys(graph).length;
+  while (queue.size() > 0) {
+    const joinOrderElement = queue.dequeue();
+    // condition to prune this branch
+    if (joinOrderElement && joinOrderElement.getCost() >= bestCost) {
+      continue;
+    }
+    // condition to stop this branch
+    if (joinOrderElement && joinOrderElement.getSize() === JOIN_ORDER_SIZE) {
+      if (joinOrderElement.getCost() < bestCost) {
+        bestCost = joinOrderElement.getCost();
+        bestJoinTables = joinOrderElement.joinTables;
+        bestJoinOrder = joinOrderElement.joinOrder;
+      }
+      continue;
+    }
+    const children = joinOrderElement ? joinOrderElement.getChildren() : [];
+    for (const child of children) {
+      queue.enqueue(child);
+    }
+  }
+  console.log('Best join order: ', bestJoinTables, ' best cost: ', bestCost);
+  return getRelationalExpression(graph, bestJoinOrder, globalSelections);
+};
+
+/**
+ * Gets the rows and cost for the given join conditions
+ * @param rows - the rows
+ * @param cost - the cost
+ * @param joinConditions - the join conditions
+ */
+const getRowsAndCost = (
+  rows: Array<any>,
+  cost: number,
+  joinConditions: Array<JoinCondition>,
+  leftTables: Array<string>,
+  rightTable: string
+) => {
```

Comment:
It should be possible to reuse the existing logic for executing queries to determine the estimated number of rows. See `applyItem` in `src/modules/data.js`. It would be great if this logic didn't need to be implemented twice.

Reply:
Tried changing `getRowsAndCost` to use `applyItem` but could not get it to work. `applyItem` expects slightly different arguments than I can provide here. Can you take a look at this once you have time ?

Reply:
What specifically did you try? I can't really identify what the problem is without seeing your code. We can discuss this when we meet.

Reply:
This is now resolved. Resued `applyExpr` from `data.js`